tech_reports/LLMs/llms.md

```diff
@@ -43,8 +43,154 @@ Authors:
   - Iterative update system
   - When to use our fused op
 ### 2.3 Norm
-  - Replicated layernorm vs distributed layernorm
-    - Layernorm/rmsnorm weights in row major / wrapped around tile size trick
+
+Normalization is a critical operation in Large Language Models (LLMs), ensuring stable training and efficient inference. Two widely adopted normalization techniques in modern LLMs, **LayerNorm** and **RMSNorm**, are fully supported in TT-NN.
+
+#### Implementations of Normalization Operations
+
+TT-NN includes two primary implementations of normalization operations to handle diverse activation layouts efficiently:
+
+1. **Non-Distributed Norm**
+2. **Distributed Norm**
+
+
+#### 1. Non-Distributed Norm
+
+This implementation supports both sharded and interleaved inputs. It is employed in the following scenarios:
+- **Single-Device Activations**: When the entire embedding resides on a single device.
+- **Multi-Device Replicated Activations**: When activation data is replicated across devices in a data-parallel setup.
+
+
+
+**Example: RMSNorm on Single Device (Decode Scenario)**
+
+```python
+import torch
+import ttnn
+
+def torch_rms_norm(x, gamma, eps):
+    return x * torch.rsqrt(x.pow(2).mean(-1, keepdim=True) + eps) * gamma
+
+batch, seq_len, embedding_dim = 32, 1, 8192
+torch_input = torch.randn((batch, seq_len, embedding_dim))
+torch_gamma = torch.randn((embedding_dim))
+torch_output = torch_rms_norm(torch_input, torch_gamma, eps=1e-5)
+
+# Reshape inputs/weights to 4D tensors
+torch_input = torch_input.view(1, 1, batch, embedding_dim)  # seq_len = 1 for decode
+torch_gamma = torch_gamma.view(1, 1, 1, embedding_dim)
+
+# Convert tensors to TT-NN tensors
+ttnn_input = ttnn.as_tensor(
+    torch_input,
+    device=device,
+    dtype=ttnn.bfloat16,
+    layout=ttnn.TILE_LAYOUT,
+    memory_config=ttnn.DRAM_MEMORY_CONFIG
+)
+ttnn_gamma = ttnn.as_tensor(
+    torch_gamma,
+    device=device,
+    dtype=ttnn.bfloat16,
+    layout=ttnn.TILE_LAYOUT,
+    memory_config=ttnn.DRAM_MEMORY_CONFIG
+)
+
+# Perform RMSNorm
+ttnn_output = ttnn.rms_norm(ttnn_input, epsilon=1e-5, weight=ttnn_gamma)
+```
+
+**Optimization for Efficient Weight Reads from DRAM**
+
+In above example, weights were traditionally pushed to device in **TILE layout**. But in this case, padding is required to match the TILE_HEIGHT. This padding increased memory footprint and reduced DRAM access efficiency. To address this, weights are now wrapped into **TILE_WIDTH** sticks and converted to **ROW_MAJOR_LAYOUT** without requiring any padding.
```

Comment:
Is there a perf cost for this transformation? Maybe mention that.

Reply:
There is no perf cost at runtime as its done once during model initialization. Added that